enhancements/ingress/set-delete-http-headers.md

```diff
@@ -0,0 +1,1031 @@
+To quote using single-quotes, we will do the following:
+1. Replace `'` with `'\''`.
+2. Wrap the entire string in single quotes.
+For example, `"abc"\\ 'def'`  will be converted to `'"abc"\\ '\''def'\'''`.
+
+
+### Risks and Mitigations
+
+We identified the following headers that require special explanations:
+
+#### Proxy header
+The `proxy` HTTP request header can be used to exploit vulnerable CGI applications using the [_httpoxy_](https://httpoxy.org/) vulnerability.
+In this vulnerability, if the client sets the HTTP `proxy` request header and the proxy passes it through to the application, then the CGI application takes the header value and injects it in the `HTTP_PROXY` environment variable.
+
+In this manner, an attacker could direct a vulnerable application to use a proxy of the attacker's choice.
+Note that OpenShift router already mitigates this vulnerability by deleting any `proxy` header that the client specifies.
+
+Allowing the cluster-admin or route owner to override the `proxy` HTTP header isn't necessarily a security risk; it would only allow admins or route owners to inject configuration into their own applications, which they already control.  However, the `proxy` header is non-standard, and allowing it to be configured is potentially
+confusing.  For this reason, we do not allow setting the `proxy` header with this EP.
+
+#### Strict-Transport-Security
+Strict-transport-security is already handled separately, and we don't want to
+allow overrides of that configuration.
+
+If we allowed it, then route owners could impact each other's routes.
+
+It must be blocked so, we have put this
+header in the deny list.
+
+#### X-Forwarded-For
+`x-forwarded-for` and related headers were considered for the deny-list (as modifying these headers using this new API could affect configuration specified using
+another [option](https://docs.openshift.com/container-platform/4.12/networking/ingress-operator.html#nw-using-ingress-forwarded_configuring-ingress))
+but have actual use-cases (which aren't covered by the existing option). They will be allowed, with notes in the API and product documentation,
+and possibly a warning on admission.
+
+For example: If someone creates a Ingress Controller with the following:
+```sh
+apiVersion: operator.openshift.io/v1
+kind: IngressController
+metadata:
+  name: default
+  namespace: openshift-ingress-operator
+spec:
+  httpHeaders:
+    forwardedHeaderPolicy: Replace
+```
+Then the `set-header` stanzas will be added in `haproxy.config` using the [haproxy-config.template](https://github.com/openshift/router/blob/0cd9a58dff77127562f7851ccfa5cda79b8cef48/images/router/haproxy/conf/haproxy-config.template#L565-L580)
+However, if someone wants to set custom values for these then they need to use the API of this EP.
+
+If both are set then custom values will take over.  
```

Comment:
There is also an annotation on the Route:  `haproxy.router.openshift.io/set-forwarded-headers`

Can you please list the four different ways `x-forwarded-for` can be specified, and how these ways interact in processing?

Reply:
A table would be excellent.  You could list operations too, to specify that deleting is only allowed by this new feature.

Reply:
We talked about this in the setheaders slack channel.

Reply:
done